### PR TITLE
Added new sanity check tables and initial values

### PR DESCRIPTION
**Pull request by Jambrose777**
New tables for the db-sanity-check project to pull from instead of storing sanity checks and exclusions to the repo itself. :D

### CODE REVIEW

File: WcaOnRails/db/migrate/20201020193829_create_wrt_sanity_check_tables.rb [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    end
+    create_table :sanity_checks do |t|
+      t.integer :sanity_check_category_id
+      t.string :topic
+      t.text :comments
+      t.text :query
+    end
+    create_table :sanity_check_exclusions do |t|
+      t.integer :sanity_check_id
+      t.json :exclusion
```
Comment (Jambrose777) [inline WcaOnRails/db/migrate/20201020193829_create_wrt_sanity_check_tables.rb:16]: hmm, this works fine locally, but seems to cause a build issue :(

**Comment by campos20** [comment]
If I'm not mistaken, the [json column](https://travis-ci.org/github/thewca/worldcubeassociation.org/builds/737549381#L1483) is triggering an issue.

**Comment by campos20** [comment]
Quick suggestion: replace with two columns: one for alias name exclusion and the other for the value.

**Comment by SAuroux** [comment]
> Quick suggestion: replace with two columns: one for alias name exclusion and the other for the value.

No, then we would need to add multiple rows per exclusion and wouldn't have a primary key anymore. Also there must be a way to get this to work, since we definitely have columns with JSON-values in other tables. 

@jonatanklosko Please advise.

**Comment by moralsh** [comment]
Why don't you just declare it to be varchar or longtext?

**Comment by Jambrose777** [comment]
> Why don't you just declare it to be varchar or longtext?

json type will actually validate that what the data is is a json. Inserting something that isnt correct json syntax would result in an sql error. Again, it worked fine locally, I'm not too sure why it causes a build failure.

**Comment by jonatanklosko** [comment]
With our current Travis configuration the builds run on `ubuntu 14` and from their docs it looks like this implies `mysql 5.6`, which indeed doesn't support the `JSON` column type (introduced in `mysql 5.7`). Please try changing [this](https://github.com/thewca/worldcubeassociation.org/blob/6b39c1216d8bcadeb28c3dd951e43b1292aac2fb/.travis.yml#L1) to `dist: focal` and pushing that change. This should bring `myslq 8.0` to match our production environment and hopefully won't trigger any other issues ^^

**Comment by jonatanklosko** [comment]
I may have missed some points, but a top level question: why do we need to introduce these tables here instead of doing so in the sanity checks repo? I assume this could easily be executed as a post-db-import script there. My main issue here is all this hard-coded data in the migration, which I assume is gonna be updated in the future.

**Comment by Jambrose777** [comment]
> I may have missed some points, but a top level question: why do we need to introduce these tables here instead of doing so in the sanity checks repo? I assume this could easily be executed as a post-db-import script there. My main issue here is all this hard-coded data in the migration, which I assume is gonna be updated in the future.

Can you bring that up in the Sanity Check email thread? I don't know all the reasons, but I guess a big one is security.

Yes, what is in the tables will update over time, probably with direct inserts to the db since there are no rails validations run. What's inserted is just where it starts with the sanity checks and exclusions not based on specific persons. Should these be left out of the migration?

**Comment by jonatanklosko** [comment]
> Can you bring that up in the Sanity Check email thread?

Sure!

> What's inserted is just where it starts with the sanity checks and exclusions not based on specific persons. Should these be left out of the migration?

If we go this path with manual inserts, then I believe it would make more sense to insert the initial data manually as well.

**Comment by jonatanklosko** [comment]
I see the build fails saying there's no database available. I see we don't actually declare we need one anywhere, kinda surprising it worked before ^^ Please try adding the following below `os: linux`:

```yaml
services:
  - mysql
```

**Comment by jonatanklosko** [comment]
Alright, so the database works fine in Travis, the remaining issues have to do with the actual code.

Firstly, you need to define how the new tables will be treated when creating a database dump. We should definitely just skip their content, same as [here](https://github.com/thewca/worldcubeassociation.org/blob/96683dacf0e7b795d5c2a8a71e172a2e0ae7839d/WcaOnRails/lib/database_dumper.rb#L445).

Then the generated `structure.sql` doesn't specify the right character encoding, which usually has to do with local database configuration. I think the simplest way is to manually fix `structure.sql`, so that Travis (and production) is happy. Something like this should do:

```sql
DROP TABLE IF EXISTS `sanity_check_categories`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `sanity_check_categories` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `name` varchar(191) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;
DROP TABLE IF EXISTS `sanity_check_exclusions`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `sanity_check_exclusions` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `sanity_check_id` int(11) DEFAULT NULL,
  `exclusion` json DEFAULT NULL,
  `comments` text CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;
DROP TABLE IF EXISTS `sanity_checks`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `sanity_checks` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `sanity_check_category_id` int(11) DEFAULT NULL,
  `topic` varchar(191) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `comments` text CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci,
  `query` text CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;
```

**Comment by Jambrose777** [comment]
Thanks @jonatanklosko  Those changes worked. Sorry it took a hot minute for me to get done.

**Comment by jonatanklosko** [comment]
I think it's alright now, @campos20 does everything look correct to you?

**Comment by campos20** [comment]
@SAuroux for this to work properly without server changes,  the query would need to be self contained.  That is

exclusion: {column: "personId", values: ["2000AAAA02", "2000BBBB03"]}

And the query stored would be something like
select * from Persons where id not it (select values from exclusion where id=1)

Adapted to json, course

**Comment by SAuroux** [comment]
@campos20 No! We had concluded this differently in October, please re-check the email discussion leading up to this PR. 

Your example would be two rows in the exclusion table and the values of the exclusion column might look like this:
{"personId": "2000AAAA02", "value2": 500}
{"personId": "2000BBBB03", "eventId": "pyram"}

I asked you on October 19th if you see a problem with this from implementation perspective and you replied with 

> No issues with that, it's an easy implementation.

I also don't understand what you mean with server changes now.

**Comment by campos20** [comment]
> I also don't understand what you mean with server changes now.

I should have mentioned _further_ server changes, meaning, no changes after we conclude this step of using database to store the queries and exclusions. I've been considering 2 options, both of them fit on this PR so, concerning changes introduced by Jacob, they are perfect. What sanity check project needs is actually just a query in `sanity_checks` and a json in `sanity_check_exclusions`. I think it would be nice if `sanity_check` and `sanity_check_exclusions` had a foreign key to `sanity_check_categories`. @Jambrose777. Also, WRT maybe would like to use the old identifiers (N1, N2...) somehow. 

What follows is not exactly related to this PR, it could be moved to sanity_check channel or so, but I see no problem in sharing this. These are the two changes I've been considering (no changes to this PR)

Option 1: Server handles exclusion. The query remains exactly the same. The json in the exclusion would need to identify which column to look for in the exclusion, the values, and the batch would handle it (pretty much what was discussed in October).

Option 2: Exclusion handled in the query, meaning WRT would have total control of it. Here's an example.

1. Run the script that creates the tables introduced by this PR
2. Populate the exclusion table with ```INSERT INTO sanity_check_exclusions (id, sanity_check_id, exclusion, comments) VALUES(1, 1, '{"wcaId": ["2008DANG02", "2008DEBR01", "2013DALM01"]}', 'Lower case, 1 wcaId missing just because')```
3. Run the modified N3 query ```SELECT
    *
FROM
    Persons p
WHERE
    BINARY MID(REVERSE(name), LOCATE(' ', REVERSE(name)) -1, 1) <> UPPER(
        MID(REVERSE(name), LOCATE(' ', REVERSE(name)) -1, 1)
    ) -- Exclusion
    and not json_contains(
        (
            select
                json_extract(exclusion, "$.wcaId")
            from
                sanity_check_exclusions sce
            where
                id = 1
        ),
        json_quote(p.id)
    )```

In both cases, a query would need to be added to the table `sanity_checks`. Either the original query or the modified one.

**Comment by SAuroux** [comment]
So we already agree that everything is OK with this PR, great! For everything else I can just reiterate: according to my understanding, we have already discussed all of this in October.

Here is what Jacob proposed and what you agreed to after some more elaboration in the email conversation: 

> For the sanity_check_exclusions, Is the value data best stored as a JSON.stringify representation of the data objection that is to be excluded?
> [...] And then the db-sanity-check would use this to identify exclusions regardless of the data used to interpret the exclusion.

This means that from your two options above, we already agreed on options 1 and we also explicitly agreed even earlier that modifying the sanity check queries to handle exclusions would not be necessary. 
It is further not correct that WRT would only have total control of exclusions with option 2. We would have that in both scenarios by making the needed changed to the sanity check tables.

> Also, WRT maybe would like to use the old identifiers (N1, N2...) somehow.

Yes, but just like now, these will just be part of _sanity_checks.topic_.

> Populate the exclusion table with INSERT INTO sanity_check_exclusions (id, sanity_check_id, exclusion, comments) VALUES(1, 1, '{"wcaId": ["2008DANG02", "2008DEBR01", "2013DALM01"]}', 'Lower case, 1 wcaId missing just because')

{"wcaId": ["2008DANG02", "2008DEBR01", "2013DALM01"]} would not be a valid exclusion, that's what I meant to point out in my last message. One row in the sanity_check_exclusions table will always correspond to exactly one to be excluded query result.

**Comment by jonatanklosko** [comment]
Should I go ahead and deploy the PR?

**Review by campos20 — APPROVED** [review]
> Should I go ahead and deploy the PR?

Yes, please

**Comment by jonatanklosko** [comment]
Deployed and migrated!

**Comment by Jambrose777** [comment]
I've gone ahead and inserted the data into the 3 tables (I'll follow up with that file in an email to WRT and Alexandre). There is a small issue:
When Accessing the sanity_check_exclusions table in phpMyAdmin I get the following error:
image

**Pull request merged**